README.md
# [Plex Optimizer](https://github.com/DavidRodriguezSoaresCUI/PlexOptimizer)

This software aims at helping you with your movies/series library of video files by making them as **widely compatible** as possible.

This software was originally developed for low-power Plex servers that can't handle real-time transcoding, by producing "high compatibility" versions of video files in advance that can be streamed without transcoding to pretty much any device, but you may find it other uses !

> You may share your use case by opening an issue with title beginning with `[USECASE]`

## Quick start guide

1. Install <a href="#requirements">Requirements</a>
2. Execute from terminal with syntax:
    ```bash
    <PYTHON> <PLEX_OPTIMIZER_PATH> <ARGUMENTS>
    ```

## Contents of this repository

I am the exclusive owner of the content of this repository, unless explicitly stated.

Required to execute `Plex Optimizer`:
- `plex_optimizer.py`: the main script
- `codec_constraint_converter.py`: bespoke module for rule-based conversion planning with ffmpeg and ffprobe
- `os_detect.py`: taken from [scivision's pybashutils](https://github.com/scivision/pybashutils/blob/main/pybashutils/os_detect.py)
- `utils.py`: contains functions and stuff, mostly taken from my other project [DRSlib](https://github.com/DavidRodriguezSoaresCUI/DRSlib)
- `webvtt_sanitize.py`: bespoke script for a subtitle conversion step
- `format_compatibility.json`: contains information about MP4 and MKV containers' format compatibilities

Repository-specific:
- `CHANGELOG.md`: for changes and version information
- `README.md`: contains this text
- `LICENSE`: for copyright information
- `ELSE`: directory containing information and PoC used during development (some contents are owned by the `ffmpeg` or `mkvmerge` projects)
- `FFMPEGContainerTester`: directory containing related project made to determine container codec compatibility using FFMPEG, used during development

## Requirements

**WARNING**: For now, it was only tested on **Windows** and **WSL**. Full Linux support is pending.

- `Python` (version >=3.6) must be installed in your system to run the script.
  - see [official site](https://www.python.org/downloads/)

- `ffmpeg`, `ffprobe`: must be installed in your system and callable from the command line.
  - **Windows**: [Latest releases](https://github.com/GyanD/codexffmpeg/releases/tag/5.0.1) (if you don't know which one to get, download ``ffmpeg-*-full_build-shared.zip``)
  - Others: see [official site](https://ffmpeg.org/download.html)

- `mkvmerge`: Needed for proper MKV output because `ffmpeg` doesn't update metadata properly

- (Optional) [PgsToSrt](https://github.com/Tentacule/PgsToSrt): used to convert image-based `PGS` subtitles (found in Blu-Ray discs) to text

You can easily find guides on how to install these dependencies on your favorite search engine.

## Help

The help contains important information about how the `Plex Optimizer` script works and the implications of using it, so it's **highly recommended** to give it a read.

You can view the help with by typing `python plex_optimizer.py  -help` or `python3 plex_optimizer.py --help` into a terminal at the location of file `plex_optimizer.py`.

```
usage: Plex Optimizer [-h] [--mode {lite,full,standalone}] [--just_one]
                      [--single_script] [--threads THREADS]
                      [--format [FORMAT ...]] [--bitrate_limit BITRATE_LIMIT]
                      [--x264_preset X264_PRESET]
                      [--x264_crf X264_CRF | --x264_target_bitrate X264_TARGET_BITRATE]
                      [DIR]

 This script tries to optimize video files for streaming playback on Plex or other
services, mainly by avoiding video and audio transcoding.

This is achieved by:
1.Converting video formats that are not h264, with bit depth other than 8-bit or
  with resolution superior to 1080p.
2.Converting non-widely-supported formats (vorbis, opus, flac, dts) to AAC if
  mono/stereo or ac3 for multi-channel
3.Converting text subtitle formats that need to be "burned into" the video stream
  to a more basic text format.
4.(Possibly unnecessary) Using MP4 container for output when possible (MKV otherwise)

Note on feature `mode`:
>'lite': Produce an optimized file as a "high compatibility" version of the original,
 so replacing "low compatibility" streams. These files are meant to coexist with
 source files as an alternative. Recommended for most and default. Output should
 always be MP4. Warning: drops image-based subtitle streams (VobSub,PGS).
>'standalone': Same as 'full' except it will not include "low compatibility" video
 streams. These files are meant to replace source files (if you don't care about
 potentially losing the original video stream). Use it if you absolutely don't want
 to have two versions of the same video or really need the disk space. Output is MP4
 if possible but usually MKV.
>'full': [WARNING: THIS IS AN EXPERIMENTAL FEATURE] Produce a file that contains
 all streams from the original file, plus the "high compatibility" ones. These files
 are meant to replace source files as a "superset" of its streams. NOT RECOMMENDED
 because video files with multiple video streams aren't supported on most players.
 Output is typically MKV.

Requirements: Python >= 3.6, also ffmpeg and ffprobe (need to be callable by their name)

WARNING:
>Any lossy media conversion involves some information/quality loss. Keep the original
 files if possible. THE AUTHOR OF THIS PROGRAM ISN'T RESPONSIBLE FOR LOSS OF DATA.
>Dealing with subtitle streams is particularly hard because of the variety of format,
 difference between image-based and text-based formats, and the lossy nature of conversion
 between some of them. There is a clear compromise between compatibility and capability
 when it comes to subtitle codecs, so keep in mind that converted "high compatibility"
 subtitles ARE NOT EQUAL to the original and MAY BE UNUSABLE.

positional arguments:
  DIR                   (Optional) Directory containing video files to
                        process.

options:
  -h, --help            show this help message and exit
  --mode {lite,full,standalone}
                        See above for which option to choose. Default: lite
  --just_one            Only process one file.
  --single_script       Only produce one script for all optimizations.
  --threads THREADS     Sets limit on threads used by h264 encoder.
  --format [FORMAT ...]
                        Override output format range (Default equivalent to
                        `--format mp4 mkv`).
  --bitrate_limit BITRATE_LIMIT
                        Maximum bitrate for a stream (default: 7M=7000000).
                        Heavier streams are forcibly converted. Accepted
                        suffixes are K (kbps) and M (mbps) (case insensitive).
  --x264_preset X264_PRESET
                        Sets value for `-preset` used by h264 encoder.
  --x264_crf X264_CRF   Sets value for `-crf` used by h264 encoder (1-pass
                        mode).
  --x264_target_bitrate X264_TARGET_BITRATE
                        Sets value for `-b` used by h264 encoder (2-pass
                        mode).

If no DIR is given, you will be prompted at runtime.
```

**Note**: default is equivalent to ```python plex_optimizer.py --mode lite```

## Usage examples

For the following examples, it is assumed that the terminal's Current Working Directory contains the `Plex Optimizer` script.

- You want to produce a high compatibility version of file `D:\movies\avatar (2009).avi` and choose mode `lite`. First, you move the file to `D:\convert` then you can run :

    ```bash
    python plex_optimizer_2.py --mode lite "D:\convert"
    ```

    It tells you you can run script `D:\convert\avatar (2009).optimizer_lite.bat` for encoding, so you run :
    ```bash
    "D:\convert\avatar (2009).optimizer_lite.bat"
    ```
    This may take a few minutes/hours, but a "high compatibility" version this movie will be produced if needed.

- You want to produce a high compatibility version of files of a show located at `D:\shows\Steven Universe`, choose mode `standalone`. You do a test run first on one file :

    ```bash
    python plex_optimizer_2.py --mode standalone --just_one "D:\shows\Steven Universe"
    ```

    It tells you you can run script `D:\shows\Steven Universe\S01E01.optimizer_standalone.bat` for encoding, so you run :
    ```bash
    "D:\shows\Steven Universe\S01E01.optimizer_standalone.bat"
    ```

    It produces file `D:\shows\Steven Universe\S01E01.optimized_standalone.mkv`. You are satisfied with the result, so you run :

    ```bash
    python plex_optimizer_2.py --mode standalone --just_one "D:\shows\Steven Universe"
    ```

    It tells you you can run script `D:\shows\Steven Universe\optimize_all.bat` for encoding, so you run :
    ```bash
    "D:\shows\Steven Universe\optimize_all.bat"
    ```
    This may take a few minutes/hours, but a "high compatibility" version of any file found in `D:\shows\Steven Universe` that can be optimized will be produced.

## To-do list

- [High priority, likely to happen] Make it work on Linux systems (mostly requires a dedicated script formatter)

- [Med priority, unlikely to happen] Implement a custom subtitle conversion process for image-based subtitle streams (would involve OCR tools like Tesseract)
  > Using project [PgsToSrt](https://github.com/Tentacule/PgsToSrt) to convert PGS subtitles (SUP container) to SRT since version `0.0.dev1`.

- [Low priority, unlikely to happen] Implement a custom subtitle conversion process for text-based subtitle streams
  > Partially done; by finding better conversion steps and implementing a custom conversion step (see changelog for version `0.0.dev3`)

## Q & A

**Q: Why do I need to move a video file I want to optimize into an empty folder ?**

`Plex Optimizer` is designed for mass conversion, so it operates on folders. It explores the content and processes any video file it finds. If you want to process a single/few file(s), you must place it/them in its/their own folder. This is temporary and you can move them back afterward.

**Q: Plex Optimizer doesn't produce optimized versions for some files. Why ?**

A: This can be caused by a number of reasons, mainly:

- `Plex Optimizer` determined there was no need for a "high compatibility" version to be produced

- `Plex Optimizer` didn't find the files

- `ffprobe` couldn't read the files

- Something went wrong (in this case you may request help from the developer or post an issue)


**Q: The optimized version can't be read by my player/requires real-time transcoding during playback. What can I do ?**


CHANGELOG.md
# Changelog for Plex Optimizer
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- Sections should be one of: Added, Changed, Fixed, Removed -->


## [0.0.dev4] - 2022-07-01
_Note: issue of styled subtitles losing their font: Unfortunately I could not find information on supported font attachment formats for the MKV container so for now only `ttf` fonts are considered (`otf` may be supported also, requires investigation)._
### Added
- Streams of type `attachment` (format `ttf` only for now) are now copied in modes `standalone` and `full`. This was done to preserve look of original subtitle streams with font support. Note: forces use of MKV container
- `SUBTITLE_FORMAT_INVESTIGATION`: Summary of intel about subtitle formats
- CLI argument `--bitrate_limit` now accepts "human-readable" suffixes (eg: `12M` equivalent to `12000000` and `650k` equivalent to `650000`)


## [0.0.dev3] - 2022-06-24
_Note: issue of lost concurrent subtitles when converting to the MP4 container (MOV_TEXT format): While the introduced solution can have ugly results (when there are a lot of concurrent subtitles or some are very long), it remains an improvement in my opinion._
### Added
- `webvtt_sanitize.py` and `webvtt_sanitize_LR()`: A step was added for the (text-based subtitle) -> MOV_TEXT conversion because the WebVtt -> MOV_TEXT conversion had issues (can't deal with blank lines in subtitle text or with concurrent subtitles). See notes in `webvtt_sanitize.py` for more information.


## [0.0.dev2] - 2022-05-22
_Note: Motivation for adding 2-pass variable bitrate mode on h264 video encoding: Users may want to target a specific bitrate, or (indirectly) a specific file size._
### Added
- CLI arguments to control common libx264 parameters: `--x264_preset` (default is `slow`), and rate mode: either CRF with `--x264_crf` (default option; default value is `22`) or variable bitrate with `--x264_target_bitrate`
- Implementation of 1-pass (CRF) vs 2-pass for h264 video encoding
- 2-pass-specific temporary file cleanup

### Changes
- Fixed issue with streams to be copied would be included in temporary files
- Various other fixes and checks


## [0.0.dev1] - 2022-05-20
_Note: While this iteration is the first to be documented in this document, it does not qualify as a *release* by decision of the author. Reasoning hinges on both the willingness to document evolution of the current project and the unwillingness to produce releases due to the current "alpha" nature of the software._

_The author plans to start producing releases when this project reaches a more
stable state._

### Added
- This CHANGELOG file to hopefully document changes, improvement, bugfixes
  and more.
- Temporary directory creation and cleanup after finishing moved to optimization
  script because it made more sense to do it that way.
- Handling of `hdmv_pgs_subtitle` subtitles in function `optimize_subtitle_to_mov_text_or_srt` for format `mkv`.
- CLI argument `--format` to override default.

### Fixed
- Constant `CodecConstraintConverter.FFMPEG_CALL` was being modified in function
  `CodecConstraintConverter.remux`, leading to hard-to-debug issues and cascading
  command corruption.
- Function `CodecConstraintConverter.plan_conversion` was ignoring `output_format`
  argument and defaulting to `mp4`.
- CLI argument `--just_one` implementation is now functional.

### Changed
- User is now prompted to choose between recursive and non-recursive file gathering
  instead of defaulting to recursive.
- `display_status` now displays status of CLI arguments `--format` and `--single_script`
- Minor CLI improvements

ELSE/plex_optimizer_OLD.py
#pylint: disable=no-name-in-module, import-error, invalid-name, wrong-import-position, global-statement
''' This script tries to optimize video files for streaming playback on Plex or other
services, mainly by avoiding video and audio transcoding.

This is achieved by:
1. Converting video formats that are not h264, with bit depth other than 8-bit or
   with resolution superior to 1080p.
2. Converting non-widely-supported formats (vorbis, opus, flac, dts) to AAC if
   mono/stereo or ac3 for multi-channel
3. Converting text subtitle formats that need to be "burned into" the video stream
   to a more basic text format.
4. Using MP4 container for output when possible (MKV otherwise)

Note on experimental feature `mode`:
    >'lite': Produce an optimized file as a "high compatibility" version of the original,
      so replacing "low compatibility" streams. These files are meant to coexist with
      source files as an alternative. Recommended for most and default. Output should
      always be MP4. Warning: drops image-based subtitle streams (VobSub,PGS).
    >'full': Produce a file that contains all streams from the original file, plus the
      "high compatibility" ones. These files are meant to replace source files as a
      "superset" of its streams. NOT RECOMMENDED because video files with multiple video
      streams aren't supported on most players. Output is MP4 if possible but typically
      MKV. WARNING: this is an experimental feature
    >'standalone': Same as 'full' except it will not include "low compatibility" video
      streams. These files are meant to replace source files (if you don't care about
      potentially losing the original video stream). Use it if you absolutely don't want
      to have two versions of the same video or really need the disk space. Output is MP4
      if possible but usually MKV.

Requirements: Python >= 3.6, also ffmpeg and ffprobe (need to be callable by their name)

Known issues:
    >When converting, ffmpeg doesn't update some metadata, so inspecting output file with
     tools like MediaInfo may contain inaccurate information. This can be solved by re-muxing
     output files with tools like Mp4Box or MKVmerge.
    >Converting ASS subtitles can result in messy output: duplicate entries, one-frame
    entries, issues with concurrent entries, etc. Dealing with subtitles automatically
    with ffmpeg is hard in general.

WARNING:
    >Any lossy media conversion involves some information/quality loss. Keep the original
     files if possible. The author of this program isn't responsible for loss of data.
    >Dealing with subtitle streams is particularly hard because of the variety of format,
     difference between image-based and text-based formats, and the lossy nature of conversion
     between some of them. There is a clear compromise between compatibility and capability
     when it comes to subtitle codecs, so keep in mind that converted "high compatibility"
     subtitles ARE NOT EQUAL to the original and MAY BE UNUSABLE.
'''


EXPLANATION = '''\
                    | 'lite' mode | 'standalone' mode |   'full' mode    |
--------------------|-------------|-------------------|------------------|
'passthrough' codec |    copy     |       copy        |      copy        |
--------------------|-------------|-------------------|------------------|
'convert' codec     |   convert   | *copy AND convert | copy AND convert |
--------------------|-------------|-------------------|------------------|
unknown codec       |    drop     |       copy        |      copy        |
--------------------|-------------|-------------------|------------------|

* For video stream, only convert
'''

FIX_LOG = '''\
# FIX LOG

## Fix 01

### Problem

When converting subtitles (ASS->SRT), the output font size information may be applied
differently between players, so some may display text too small or large. There appear
to be no fix for this because it's how ffmpeg does the conversion. Other artifacts like
broken or missing text is to be expected. Worse, a report of Plex Smart TV app freezing
while trying to render such subtitles raised the importance of the issue.

### Fix description

- Idea 01: Using 'text' as encoder instead of 'srt' removes all markup tags. This may not be desirable,
as it also removes effects such as making the text bold or italic.

- Idea 02: It was discovered that when converting ASS subtitles to WebVTT first, then to SRT, the
<font ..> tag is discarded while bold and italic tags are kept.

- Idea 03: It was discovered that when converting ASS subtitles to 'mov_text', the
<font ..> tag is discarded while bold and italic tags are kept. Also, this codec
is compatible with the MP4 container.

### Decision

Idea 02 was selected and is pending implementation, with idea 01 was selected as alternative.
'''

FEATURE_IDEAS = '''\
> Convert image subtitles with tools based on Tesseract (like https://github.com/Tentacule/PgsToSrt)
'''

FULL_MODE_WARNING = '''\
WARNING: you chose `full` mode. Please read notes regarding this mode in the help
by running `plex_optimizer` with parameter `--help`. You must understand that this
feature is experimental and that video player support for output files is limited.

Press ENTER to continue or CRTL+C to quit
'''

HONORABLE_MENTIONS = '''\
The following links were insightful when crafting this script:
https://stackoverflow.com/questions/54659242/ffmpeg-convert-all-audio-streams-from-mkv-to-mp4
https://stackoverflow.com/questions/43727447/ffmpeg-recode-all-audio-streams-while-keeping-originals
https://stackoverflow.com/questions/32922226/extract-every-audio-and-subtitles-from-a-video-with-ffmpeg
https://support.plex.tv/articles/200471133-adding-local-subtitles-to-your-media/
'''

import argparse
import json
from collections import namedtuple
from pathlib import Path
from typing import List

from DRSlib.cli_ui import cli_explorer
from DRSlib.path_tools import file_collector
from DRSlib.execute import execute
from DRSlib.path_tools import find_available_path


CodecRule = namedtuple(typename='CodecRule', field_names=['passthrough','convert'])
ConversionStep = namedtuple(typename='ConversionStep', field_names=['codec', 'parameters'])
VIDEO_EXT = { '.mp4', '.mkv' }
CODEC_OPTIMIZATION_RULES = {
    "video": CodecRule(
        passthrough={},
        convert={ 'mjpeg', 'mpeg4', 'mpeg2video', 'hevc', 'av1' }
    ),
    "audio": CodecRule(
        passthrough={ 'mp3', 'aac', 'ac3', 'eac3' },
        convert={ 'flac', 'vorbis', 'opus', 'dts', 'mp2' }
    ),
    "subtitle": CodecRule(
        passthrough={ 'mov_text' },
        convert={ 'ass', 'ssa', 'subrip' }
    )
}
H264_EXTRA_PARAMS = [ "-preset", "slow", "-crf", "22" ]
FFMPEG_PARAMS = "-loglevel warning -stats -probesize 100G -analyzeduration 100G"
DEBUG = False
MKV_PATCH = {
    'mov_text': 'webvtt' # Fix for error `Subtitle codec 94213 is not supported.`
}

# V: 'mjpeg', 'h264', 'mpeg4', 'mpeg2video', 'hevc', 'av1'
# A: 'aac', 'vorbis', 'ac3', 'eac3', 'mp3', 'opus', 'flac', 'dts', 'mp2'
# S: 'ass', 'mov_text', 'subrip', 'hdmv_pgs_subtitle', 'dvd_subtitle'
# WEIRD: 'png'

CWD = Path(".").resolve()
SCRIPT_PATH = Path(__file__).resolve().parent
CODEC_TYPES = set(CODEC_OPTIMIZATION_RULES.keys())
CMD_ENCODE = lambda f, mode, mp4_ffmpeg_parameters, mkv_ffmpeg_parameters: f'''\
chcp 65001
@echo off
SET IN_FILE="{f.resolve()}"
SET OUT_FILE_MP4="{f.resolve().with_suffix(f'.optimized_{mode}.mp4')}"
SET OUT_FILE_MKV="{f.resolve().with_suffix(f'.optimized_{mode}.mkv')}"
IF EXIST %IN_FILE% (
    echo Converting %IN_FILE%..
    echo Trying to encode in a MP4 container & ( ffmpeg {FFMPEG_PARAMS} -i %IN_FILE% {' '.join(mp4_ffmpeg_parameters)} %OUT_FILE_MP4% ) || ( del %OUT_FILE_MP4% & echo Attempt failed. Trying to encode in a MKV container & ffmpeg {FFMPEG_PARAMS} -i %IN_FILE% {' '.join(mkv_ffmpeg_parameters)} %OUT_FILE_MKV% ) || ( del %OUT_FILE_MKV% & echo WARNING! ATTEMPTS TO ENCODE TO MP4 AND MKV FAILED! )
) ELSE (
    echo ERROR: File %IN_FILE% not found!
)
'''


def dump( root: Path, msg: str ) -> Path:
    ''' Dump message to file
    '''
    dump_file = find_available_path(
        root=root,
        base_name='dump.tmp',
        file=True
    )
    dump_file.write_text(
        msg,
        encoding='utf8',
        errors='xmlcharrefreplace'
    )
    return dump_file


def dump_json( root: Path, obj: str ) -> Path:
    ''' Dump message to file
    '''
    dump_file = find_available_path(
        root=root,
        base_name='dump.json',
        file=True
    )
    dump_file.write_text(
        json.dumps(
            obj,
            indent=2,
            default=str
        ),
        encoding='utf8',
        errors='xmlcharrefreplace'
    )
    return dump_file


def dir_to_work_on( root_dir: Path ) -> Path:
    ''' Returns a Path representing the directory
    selected by the user. Script is terminated on
    KeyboardInterrupt.
    '''
    return cli_explorer(
        root_dir,
        allow_mkdir=False
    )


def probe( _file: Path ) -> dict:
    ''' Use ffprobe to get stream information
    '''
    # Run ffprobe
    cmd = [
        'ffprobe',
        '-loglevel', 'error', # disable most messages
        '-show_entries', 'stream', # output all streams
        '-of', 'json', # output format as json
        _file
    ]
    stdX = execute( cmd, shell=False )
    # Handle and decode output
    if stdX['stderr'] != '':
        print("Something went wrong: ffprobe stderr is:")
        print(stdX['stderr'])
        return None
    try:
        file_info = json.loads(stdX['stdout'])
        assert "streams" in file_info, "No 'streams' in info"
    except (json.JSONDecodeError, AssertionError) as e:
        print(e)
        dump_file = dump(CWD, stdX['stdout'])
        print(f"FFprobe output is invalid, see file {dump_file} for content")
        return None

    # package information
    return {
        s['index']:s
        for s in file_info['streams']
        if s['codec_type'] in CODEC_TYPES
    }


def optimize_video_to_h264( idx, stream_info ) -> ConversionStep:
    ''' Convert video stream to h264 with resolution <=FHD and
    bit depth 8.
    '''
    w,h = stream_info.get("width",0), stream_info.get("height",0)
    res = []
    if w < 1 or h < 1:
        print(f"Warning: Can't retrieve width or height for stream {idx} (video)." \
            + "Assuming file isn't broken and resolution is <=1080p.")
    elif w > 1920:
        # resolution limit
        res += [ "-vf", "scale=w=1920:h=-1"]
    if stream_info.get("pix_fmt",'')=="yuv420p10le":
        # 10-bit -> 8bit
        res.append("-pix_fmt yuv420p")
    res += H264_EXTRA_PARAMS
    return ConversionStep(codec='h264', parameters=res)


def optimize_audio_to_aac_or_ac3( idx, stream_info ) -> ConversionStep:
    ''' Convert audio stream to aac or ac3
    '''
    nb_channels = stream_info.get("channels",0)
    if nb_channels < 1:
        print(f"Warning: stream {idx} (audio) doesn't hava a channel count." \
            + "Assuming mono or stereo.")
    if nb_channels <= 2:
        return ConversionStep(codec='aac', parameters=[ "-q:{idx}", "1.4" ])
    return ConversionStep(codec='ac3', parameters=[])


def optimize_subtitle_to_srt( *_ ) -> ConversionStep:
    ''' Used for ASS -> SRT conversion '''
    return ConversionStep(codec='srt', parameters=[])


def optimize_subtitle_to_mov_text( *_ ) -> ConversionStep:
    ''' Used for ASS -> tx3g (mov_text) conversion '''
    return ConversionStep(codec='mov_text', parameters=[])


def stream_optimization_step( idx: int, stream_info: dict, mode: str ) -> dict:
    ''' Plans optimization for a particular stream
    '''
    optimizations = {
        "video": optimize_video_to_h264,
        "audio": optimize_audio_to_aac_or_ac3,
        "subtitle": optimize_subtitle_to_mov_text
    }

    codec_type, codec_name = stream_info.get("codec_type"), stream_info.get("codec_name")

    if codec_name in CODEC_OPTIMIZATION_RULES[codec_type].passthrough:
        # Passthrough => copy
        return ConversionStep(codec='copy', parameters=[])
    elif codec_name not in CODEC_OPTIMIZATION_RULES[codec_type].convert:
        # Unlisted codec => drop or copy
        return ConversionStep(codec='drop' if mode=='lite' else 'copy', parameters=[])

    # Current stream needs to be converted => construct optimization plan
    stream_optimization_plan = optimizations[codec_type](idx,stream_info)
    # Edit title
    stream_title = stream_info.get("tags",{}).get("title")
    stream_optimization_plan = ConversionStep(
        codec=stream_optimization_plan.codec,
        parameters=stream_optimization_plan.parameters + [
            '-metadata:s:{idx}',
            f'title="[COMPAT] {stream_title}"' if stream_title else 'title="[COMPAT]"'
        ]
    )

    # Print info
    stream_lang = stream_info.get("tags",{}).get("language",'?')
    print(f"Optimizing stream {idx} (lang:{stream_lang}): " \
        + f"{codec_name} -> {stream_optimization_plan.codec}")

    return stream_optimization_plan


def streams_optimization( streams_info: dict, mode: str ) -> dict:
    ''' Plans optimization for video file, stream by stream
    '''
    return {
        idx:stream_optimization_step(idx, stream_info, mode)
        for idx, stream_info in streams_info.items()
    }



def produce_optimized_video_script( _file: Path, destination_dir: Path, optimization_plan: dict, mode: str, cmd_file: Path = None ) -> dict:
    ''' Writes a BAT script that produces optimized file using ffmpeg.

    Notes on ffmpeg:
    Select input file stream with: -map 0:<input_file_stream_index> h264
    Set video codec with: -c:<output_file_stream_index> h264
    '''
    assert mode in { 'lite', 'full', 'standalone' }

    highest_stream = max(optimization_plan.keys())

    def lite_cmd():
        nonlocal optimization_plan
        _cmd = []
        out_idx = 0
        for in_idx in range(highest_stream+1):
            if in_idx not in optimization_plan or optimization_plan[in_idx].codec=='drop':
                # Marked to be dropped, or probably not video, audio or subtitle
                print(f"Stream {in_idx} was dropped.")
                continue

            # idx in optimization_plan and not dropped => deal with 'copy', '<codec>'
            _codec, _parameters = optimization_plan[in_idx].codec, optimization_plan[in_idx].parameters
            _cmd += [ "-map", f"0:{in_idx}", f"-c:{out_idx}", _codec ]
            _cmd += [ p.replace('{idx}',str(out_idx)) for p in _parameters ]
            out_idx += 1

        return _cmd

    def full_cmd():
        nonlocal optimization_plan
        _cmd = []
        out_idx = 0 # stream id for output file
        for in_idx in range(highest_stream+1):

            if in_idx not in optimization_plan or optimization_plan[in_idx].codec=='drop':
                # Marked to be dropped, or probably not video, audio or subtitle
                raise ValueError(f"Did not expect for stream {in_idx} to be dropped with mode `full`.")

            _cmd += [ "-map", f"0:{in_idx}", f"-c:{out_idx}", 'copy' ]
            out_idx += 1

            _codec, _parameters = optimization_plan[in_idx].codec, optimization_plan[in_idx].parameters
            _cmd += [ "-map", f"0:{in_idx}", f"-c:{out_idx}", _codec ]
            _cmd += [ p.replace('{idx}',str(out_idx)) for p in _parameters ]
            out_idx += 1

        return _cmd

    def standalone_cmd():
        nonlocal optimization_plan
        _cmd = []
        out_idx = 0 # stream for output file
        for in_idx in range(highest_stream+1):

            if in_idx not in optimization_plan or optimization_plan[in_idx].codec=='drop':
                # Marked to be dropped, or probably not video, audio or subtitle
                raise ValueError(f"Did not expect for stream {in_idx} to be dropped with mode `standalone`.")

            _codec, _parameters = optimization_plan[in_idx].codec, optimization_plan[in_idx].parameters
            if _codec!='h264':
                # Not a "low compatibility" video stream => copy
                _cmd += [ "-map", f"0:{in_idx}", f"-c:{out_idx}", 'copy' ]
                out_idx += 1

            _cmd += [ "-map", f"0:{in_idx}", f"-c:{out_idx}", _codec ]
            _cmd += [ p.replace('{idx}',str(out_idx)) for p in _parameters ]
            out_idx += 1

        return _cmd

    def mkv_patch( commands: List[str] ) -> str:
        ''' Generates MKV-specific command for ffmpeg, to avoid incompatible codecs
        breaking conversion '''
        for s in commands:
            if s in MKV_PATCH:
                yield MKV_PATCH[s]
            else:
                yield s


    # Get command and write worker script
    mode_switch = { 'lite':lite_cmd, 'full':full_cmd, 'standalone':standalone_cmd }
    cmd = mode_switch[mode]()
    if not cmd_file:
        cmd_file = find_available_path(
            root=destination_dir,
            base_name=_file.with_suffix(f'.optimizer_{mode}.bat').name
        )
        print(f"Run '{cmd_file.resolve()}' to produce optimized video.")
    with cmd_file.open('a',encoding='utf8') as f:
        f.write(
            CMD_ENCODE(
                f=_file.resolve(),
                mode=mode,
                mp4_ffmpeg_parameters=cmd,
                mkv_ffmpeg_parameters=list(mkv_patch(cmd))
            ) + '\n'
        )


def get_args() -> argparse.Namespace:
    ''' Returns a namespace representing command-line arguments
    '''
    parser = argparse.ArgumentParser(
        prog="Plex Optimizer",
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="If no DIR is given, you will be prompted at runtime."
    )
    parser.add_argument(
        'directory',
        metavar='DIR',
        nargs='?',
        help="(Optional) Directory containing video files to process."
    )
    parser.add_argument(
        '--mode',
        choices=['lite', 'full', 'standalone'],
        default='lite',
        help="See above for which option to choose"
    )
    parser.add_argument(
        '--just_one',
        action="store_true",
        help="Only process one file."
    )
    parser.add_argument(
        '--single_script',
        action="store_true",
        help="Only produce one script for all optimizations."
    )
    parser.add_argument(
        '--threads',
        action="store",
        help="Sets limit on threads used by h264 encoder."
    )
    return parser.parse_args()


def main():
    '''main'''

    args = get_args()
    global H264_EXTRA_PARAMS
    if args.threads:
        H264_EXTRA_PARAMS += [ '-threads', args.threads ]
    if args.mode=='full':
        input(FULL_MODE_WARNING)
    # Get source directory
    if args.directory:
        src_dir = Path(args.directory).resolve()
    if not args.directory or not src_dir.is_dir():
        try:
            base_dir = Path(input("Base path to explore: "))
            if not base_dir.is_dir():
                print("Dir not found. Using CWD..")
                base_dir = CWD

            src_dir = dir_to_work_on(base_dir)
        except KeyboardInterrupt:
            return

    # Collect files to optimize
    src_dir_files = list(
        filter(
            lambda f: f.suffix in VIDEO_EXT and '.optimized' not in f.name,
            file_collector( root=src_dir ) # recursive
        )
    )
    if len(src_dir_files)==0:
        print(f"No file found in {src_dir}!")
        return

    # Process files
    nb_files = len(src_dir_files)
    cmd_file = None
    if args.single_script:
        cmd_file = find_available_path(
            root=src_dir,
            base_name="optimize_all.bat"
        )
    for idx, f in enumerate(src_dir_files):
        print(f"[{idx+1}/{nb_files}] {f}")

        # Get file stream info
        streams_info = probe(f)

        # Optimize file
        optimization_plan = streams_optimization(streams_info, args.mode)
        if optimization_plan:
            produce_optimized_video_script(
                f,
                destination_dir=f.parent,
                optimization_plan=optimization_plan,
                mode=args.mode,
                cmd_file=cmd_file
            )

            if DEBUG:
                (f.parent / f"{f.stem}.opti_plan.json").write_text(
                    json.dumps(optimization_plan, indent=2),
                    encoding='utf8',
                    errors='ignore',
                )
            if args.just_one:
                break

    if args.single_script:
        print(f"Run '{cmd_file.resolve()}' to produce optimized videos.")


if __name__=="__main__":
    main()
    print("END OF PROGRAM")
